<a href="https://colab.research.google.com/github/sumitmsc/PythonPractice/blob/main/Reccursion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
🔁 What is Recursion in Python?

Recursion is a programming technique where a function calls itself to solve a smaller version of the same problem until it reaches a base condition (stopping point).

Think of it like a mirror reflecting another mirror — the process continues until a limit is reached.

In [4]:
def factorial(n):
    """
    Recursive function to find the factorial of a number.
    factorial(n) = n * factorial(n-1)
    Base condition: factorial(0) = 1
    """
    if n == 0 or n==1:
        return 1        # base case
    else:
        return n * factorial(n - 1)   # recursive call

print(factorial(5))
print(factorial.__doc__)


120

    Recursive function to find the factorial of a number.
    factorial(n) = n * factorial(n-1)
    Base condition: factorial(0) = 1
    


how it work
factorial(5)
= 5 * factorial(4)
= 5 * 4 * factorial(3)
= 5 * 4 * 3 * factorial(2)
= 5 * 4 * 3 * 2 * factorial(1)
= 5 * 4 * 3 * 2 * 1
= 120


In [6]:
#fabonanci Series
def fibonanci(n):
  if n==0 or n==1:
     return n
  else:
    return fibonanci(n-1)+fibonanci(n-2)


print(fibonanci(4))

3


✅ Explanation:
f(4) = f(3) + f(2)
= (f(2) + f(1)) + (f(1) + f(0))
= (1 + 1) + (1 + 0)
= 3